### PR TITLE
[4.0] [a11y] Pagination site

### PR DESCRIPTION
**Pull request by brianteeman**
The aim of this PR is to make the pagination seen in list views in the front end accessible - there are no visual changes

- Wrap the pagination in a nav and add a role (needed for IE) and an aria-label
- Ensure font icons have aria-hidden=true
- Add aria-current to the selected page
- Add aria-title to the active links eg Go to page 1

Note 1
that the disabled links will not be announced to a screen reader which is the intended behaviour

Note 2
the list is an unordered list which appears to be the standard as otherwise the links would be announced as Link 1 go to page 2 etc which is nonsense

### Changes
**libraries\src\Pagination\Pagination.php**
Updated to use the layouts as per the note that they were to be used from J4

_However the layouts were not correct and not accessible_

**layouts\joomla\pagination\list.php**
Simplified - as far as I could tell 

**layouts\joomla\pagination\link.php**
Updated for a11y - the a11y should be fine but the code could probably be improved perhaps

### CODE REVIEW

File: layouts/joomla/pagination/link.php [hunk 1]
```diff
@@ -21,26 +21,31 @@
 	// Check for "Start" item
 	case Text::_('JLIB_HTML_START') :
 		$icon = 'fa fa-angle-double-left';
+		$aria= JText::sprintf('JLIB_HTML_GOTO_POSITION', strtolower($item->text));
```
Comment (Quy) [inline layouts/joomla/pagination/link.php]: add space before `=`. repeat below.

File: layouts/joomla/pagination/link.php [hunk 2]
```diff
@@ -61,15 +66,8 @@
...
-
-	$title = '';
-
-	if (!is_numeric($item->text))
-	{
-		$cssClasses[] = 'hasTooltip';
-		$title = ' title="' . $item->text . '" ';
-	}
-
+	$class = "active";
```
Comment (Quy) [inline layouts/joomla/pagination/link.php]: change `"` to `'`

File: layouts/joomla/pagination/list.php [hunk 3]
```diff
@@ -9,21 +9,21 @@
...
-	<li class="pagination-start<?php echo $startDisabled; ?> page-item"><?php echo $list['start']['data']; ?></li>
-	<li class="pagination-prev<?php echo $prevDisabled; ?> page-item"><?php echo $list['previous']['data']; ?></li>
-	<?php foreach ($list['pages'] as $page) : ?>
-		<?php $disabled = $page['active'] ? '' : ' disabled'; ?>
-		<?php echo '<li class="page-item' . $disabled . '">' . $page['data'] . '</li>'; ?>
-	<?php endforeach; ?>
-	<li class="pagination-next<?php echo $nextDisabled; ?> page-item"><?php echo $list['next']['data']; ?></li>
-	<li class="pagination-end<?php echo $endDisabled; ?> page-item"><?php echo $list['end']['data']; ?></li>
-</ul>
+<nav role="navigation" aria-label="<?php echo Text::_('JLIB_HTML_PAGINATION'); ?>">
```
Comment (Quy) [inline layouts/joomla/pagination/list.php:17]: Is `role="navigation"` necessary? Per Bootstrap;
```
Ensure accessibility by using a <nav> element or, if using a more generic element such as a <div>, add a role="navigation" to every navbar to explicitly identify it as a landmark region for users of assistive technologies.
```

File: layouts/joomla/pagination/link.php (hunk 1 above)
Comment (brianteeman) [reply layouts/joomla/pagination/link.php]: done

File: layouts/joomla/pagination/list.php (hunk 3 above)
Comment (brianteeman) [reply layouts/joomla/pagination/list.php:17]: yes because of support for IE

**Comment by brianteeman** [comment]
> layouts\joomla\pagination\links.php
Not touched this file but I can't see it being used???

Looks like this file is used in the admin (when you turn off the override) I still dont understand why we have a different layout

**Comment by rdeutz** [comment]
We have two types of pagination and if I remember correctly it depends how you create the pagination. Its strange and don't asked me why it is like it is.

**Comment by brianteeman** [comment]
well I will make sure the other one is a11y compliant when I remove the atum override etc and after #21293 is merged

**Comment by brianteeman** [comment]
@ciar4n can you advise on the css please

**Comment by ciar4n** [comment]
```
.disabled.page-link:hover {
  background: transparent;
}
.disabled.page-link {
  color: #999;
}
.active.page-link.current {
  background-color: #dee2e6;
  color: #999;
}
```
Translates to the following in scss..
```
.page-link {
  &.active {
    background-color: #dee2e6;
    color: #999;
  }
  &.disabled {
    color: #999;
    &:hover {
      background: transparent;
    }
  }
}
```

I dont believe there is need for the `.current`

**Comment by ciar4n** [comment]
> CSS ? on the links
Currently only the actual number is clickable - it should be the entire area - please advise on this

You need to remove the padding from the `.page-link` and add it to the `.page-link a` instead

**Comment by brianteeman** [comment]
And which scss file would be the best. It should be global and not template
specific.

**Comment by ciar4n** [comment]
We currently don't have any CSS file loaded globally. Why do you say this should be global?

**Comment by dgrammatiko** [comment]
> It should be global and not template specific.

It should be layout (pagination) specific!

Also I thought that we will not depend on Font awesome for the front end but I still see FA classes and also not even loading the required asset here... 😓

**Comment by brianteeman** [comment]
I give up. You make rules that are not documented anywhere. If it's not
documented then it's not a rule.

**Comment by brianteeman** [comment]
@ciar4n maybe I used the wrong terminology. But surely we have some standard CSS that is used everywhere and can be used by any template.

**Comment by mbabker** [comment]
> Also I thought that we will not depend on Font awesome for the front end but I still see FA classes and also not even loading the required asset here...

Nothing to do with this pull request...

> But surely we have some standard CSS that is used everywhere and can be used by any template.

Nope.  We're trying to standardize CSS classes that won't change because of a framework update, but there is no standard CSS that is included into every page/template.

**Comment by dgrammatiko** [comment]
>  You make rules that are not documented anywhere.

Seriously now... You need to grow up dude...

**Comment by ciar4n** [comment]
> But surely we have some standard CSS that is used everywhere and can be used by any template.

Excluding the typical reset or normalise CSS, there is very little non-opinionated 'standard CSS' that would be used by every template. I think the best we can do is to lead by good example and place all a11y CSS in the cassiopeia template.

File: layouts/joomla/pagination/link.php [hunk 4]
```diff
@@ -21,26 +21,31 @@
 	// Check for "Start" item
 	case Text::_('JLIB_HTML_START') :
 		$icon = 'fa fa-angle-double-left';
+		$aria = JText::sprintf('JLIB_HTML_GOTO_POSITION', strtolower($item->text));
```
Comment (Quy) [inline layouts/joomla/pagination/link.php]: Replace with `Text::`

File: layouts/joomla/pagination/link.php [hunk 5]
```diff
@@ -78,13 +76,18 @@
 }
 ?>
 <?php if ($displayData['active']) : ?>
-	<li>
-		<a <?php echo $cssClasses ? 'class="' . implode(' ', $cssClasses) . '"' : ''; ?> <?php echo $title; ?> href="#" onclick="<?php echo $onClick; ?>">
+	<li class="<?php echo $class; ?> page-link">
+		<a <?php echo $cssClasses ? 'class="' . implode(' ', $cssClasses) . '"' : ''; ?> <?php echo $title; ?> aria-label="<?php echo $aria; ?>" href="#" onclick="<?php echo $onClick; ?>">
```
Comment (Quy) [inline layouts/joomla/pagination/link.php]: `$cssClasses` is always an empty array since `$cssClasses[] = 'hasTooltip';` is removed.

**Comment by brianteeman** [comment]
@ciar4n I applied the scss you recommended but it really does need the current so hopefully my scss is ok now

**Comment by brianteeman** [comment]
I included the compiled css so that it can be tested with patchtester. I assume that they wont be included when committed

**Comment by ghost** [comment]
@brianteeman can you please give Test Instructions?

**Comment by brianteeman** [comment]
original post updated

**Comment by ghost** [comment]
thanks for Update @brianteeman

**Comment by wilsonge** [comment]
Can you fix conflicts here please so I can merge this :)

**Comment by brianteeman** [comment]
@wilsonge are you commiting the compiled/minified files

File: templates/cassiopeia/css/template-rtl.css [hunk 6]
```diff
@@ -9407,7 +9422,7 @@ joomla-alert {
...
-    grid-template-areas: ". head head head head ." ". banner banner banner banner ." ". top-a top-a top-a top-a ." ". top-b top-b top-b top-b ." ". main main main main ." ". bot-a bot-a bot-a bot-a ." ". bot-b bot-b bot-b bot-b ." ". footer footer footer footer ." ". debug debug debug debug .";
+    grid-template-areas: ". head head head head ."
 ". banner banner banner banner ."
 ". top-a top-a top-a top-a ."
 ". top-b top-b top-b top-b ."
 ". main main main main ."
 ". bot-a bot-a bot-a bot-a ."
 ". bot-b bot-b bot-b bot-b ."
 ". footer footer footer footer ."
 ". debug debug debug debug .";
```
Comment (infograf768) [inline templates/cassiopeia/css/template-rtl.css:9425]: What are these weird characters appearing in white on red background?
Also below.

**Comment by brianteeman** [comment]
@wilsonge that should be conflicts fixed

File: templates/cassiopeia/css/template-rtl.css (hunk 6 above)
Comment (wilsonge) [reply templates/cassiopeia/css/template-rtl.css:9425]: It's a windows related issue with spacings when compiling SASS - we get it in Ciaran and Charlie's PRs too. It's fine for now. Something we need to figure out before stable though

**Comment by wilsonge** [comment]
For now we're committing minified template CSS - this will change in the future though once we've made a decision on where to put the template SASS files and whether they should go into the media directory or not like @dgrammatiko proposed.

**Comment by brianteeman** [comment]
ok and thanks

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**